receiver/googlecloudpubsubpushreceiver/receiver.go

```diff
@@ -5,38 +5,300 @@ package googlecloudpubsubpushreceiver // import "github.com/open-telemetry/opent
 	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
 
+	"cloud.google.com/go/storage"
+	gojson "github.com/goccy/go-json"
+	"go.opentelemetry.io/collector/client"
 	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componentstatus"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
 	"go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding"
 )
 
+const (
+	bucketIDKey         = "bucketId"
+	objectIDKey         = "objectId"
+	eventTypeKey        = "eventType"
+	eventObjectFinalize = "OBJECT_FINALIZE"
+
+	bucketMetadataKey          = "bucket"
+	objectMetadataKey          = "object"
+	subscriptionMetadataKey    = "subscription"
+	messageIDMetadataKey       = "message_id"
+	deliveryAttemptMetadataKey = "delivery_attempt"
+)
+
 type pubSubPushReceiver struct {
-	cfg *Config
+	cfg           *Config
+	settings      receiver.Settings
+	storageClient *storage.Client
+
+	server     *http.Server
+	shutdownWG sync.WaitGroup
+
+	nextLogs consumer.Logs
 }
 
 var _ receiver.Logs = (*pubSubPushReceiver)(nil)
 
-func newPubSubPushReceiver(cfg *Config) *pubSubPushReceiver {
+func newPubSubPushReceiver(
+	cfg *Config,
+	set receiver.Settings,
+	nextLogs consumer.Logs,
+) *pubSubPushReceiver {
 	return &pubSubPushReceiver{
-		cfg: cfg,
+		cfg:      cfg,
+		settings: set,
+		nextLogs: nextLogs,
 	}
 }
 
-func (p *pubSubPushReceiver) Start(_ context.Context, host component.Host) error {
-	_, errLoad := loadEncodingExtension[encoding.LogsUnmarshalerExtension](
+func addHandlerFunc[T any](
+	mux *http.ServeMux,
+	endpoint string,
+	unmarshal func([]byte) (T, error),
+	consume func(context.Context, T) error,
+	storageClient *storage.Client,
+	includeMetadata bool,
+	logger *zap.Logger,
+) {
+	mux.HandleFunc(endpoint, func(resp http.ResponseWriter, req *http.Request) {
+		handlerCtx := req.Context()
+		if err := handlePubSubPushRequest(
+			handlerCtx,
+			req.Body,
+			unmarshal,
+			consume,
+			storageClient,
+			includeMetadata,
+		); err != nil {
+			// Pub/Sub retries everything that is not a valid response. A valid response
+			// has the following HTTP codes: [102, 200, 201, 202, 204].
+			// You can verify this in the official documentation. For this, refer to
+			// https://cloud.google.com/pubsub/docs/push.
+			//
+			// This becomes an issue for permanent errors, because it means that if we
+			// don't return one of those codes, Pub/Sub will keep retrying. This would
+			// only stop retrying if:
+			// 1. We add event arc advanced after Pub/Sub. For this, we need to set the
+			// correct HTTP code for permanent/transient errors. You can refer to
+			// the official documentation for this. See:
+			// https://docs.cloud.google.com/eventarc/advanced/docs/retry-events#transient
+			// 2. Or return 2xx for permanent errors. Since this is not semantically correct,
+			// we are holding on this option.
```

Comment:
@axw Here is the comment. What do you think of it? Is it time for us to return 2xx for permanent errors? I feel torn on it. It doesn't seem right, but at the same time, it is wrong to retry permanent errors

Reply:
Yes, that's fine by me